app/scripts/lib/seed-phrase-verifier.js

```diff
@@ -0,0 +1,43 @@
+const KeyringController = require('eth-keyring-controller')
+
+const seedPhraseVerifier = {
+
+  verifyAccounts(createdAccounts, seedWords) {
+
+    return new Promise((resolve, reject) => {
+
+      if (!createdAccounts || createdAccounts.length < 1) {
+        return reject(new Error('No created accounts defined.'))
+      }
+
+      let keyringController = new KeyringController({})
+      let Keyring = keyringController.getKeyringClassForType('HD Key Tree')
+      let opts = {
+        mnemonic: seedWords,
+        numberOfAccounts: createdAccounts.length,
+      }
+
+      let keyring = new Keyring(opts)
+      keyring.getAccounts()
+        .then((restoredAccounts) => {
+
+          log.debug('Created accounts: ' + JSON.stringify(createdAccounts))
+          log.debug('Restored accounts: ' + JSON.stringify(restoredAccounts))
+
+          if (restoredAccounts.length != createdAccounts.length) {
+            // this should not happen...
+            return reject(new Error("Wrong number of accounts"))
+          }
+
+          for (let i = 0; i < restoredAccounts.length; i++) {
+            if (restoredAccounts[i] !== createdAccounts[i]) {
+              return reject(new Error('Not identical accounts! Original: ' + createdAccounts[i] + ', Restored: ' + restoredAccounts[i]))
+            }
+          }
+          return resolve()
+        })
+    })
+  }
+}
+
+module.exports = seedPhraseVerifier
```

Comment:
If this would be the point where the "this should NEVER happen" error happens, then the error message should be designed more dramatic (although I guess this would be core-teams work, to decide how to proceed here).

@danfinlay , please clarify: is this here where the major bug would be catched?

Reply:
@scsaba , your code is very clean and mostly self-explaining (for a project insider fully self-explaining). Personally I would need some notes to be clear about: one seed-phrase can recreate multiple accounts? What is an account (key-value pair? ... ).

Note that this code (beside doing detective-work to find the should-never-happen-bug), will serve a limited teach-in to this problem (e.g. for people visiting the project to spot this critical bug).

Reply:
@lazaridiscom, to my understanding, the seed words can recreate your primary keyring and the accounts belonging to it. The created accounts in the primary keyring are always the same. E.g. if you had 3 accounts and you restore your vault using your seed words, initially you will get one account and you can use the create account option to restore the other 2. They will be the same as you had before. The keyring creates the accounts in the same sequence. I'm sure @danfinlay can correct me if I'm wrong.

Reply:
@scsaba , thanks for the info. Just with "notes" I meant "within the code" (some comments)

Reply:
@danfinlay , re restore, please see: #3424